### ISSUE TITLE
Are filter expressions support on a Query?

### ISSUE THREAD

**Issue by mswolfe** [derjust/spring-data-dynamodb#42]
I need to be able to issue a query on a GSI and then further filter the result set based upon a filter expression.  I believe i should be able to create the GSI with just a Hash key or with both a Hash and Range key.

For reference, this is what I was trying to do: http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/QueryAndScan.html#FilteringResults

Thanks again for all of the help.

**Comment by mswolfe** [comment]
After performing some testing I've come to the conclusion that FilterExpressions are not supported for a Query.

Given a method name like this:
```java
// There is a GSI on tag that works successfully by itself, but I want to filter on customerId using a FilterExpression.
public CustomerHistory findByTagAndCustomerId(String, tag, String customerId)
```

There are two things stopping this feature from working.
1) The obvious issue is that there is no implementation dealing with adding the filter expression and filter expression attribute values to a query.
2) The [AbstractDynamoDBQueryCriteria.isApplicableForGlobalSecondaryIndex](https://github.com/derjust/spring-data-dynamodb/blob/master/src/main/java/org/socialsignin/spring/data/dynamodb/repository/query/AbstractDynamoDBQueryCriteria.java#L358) method returns false when property names are present that aren't in an index.

I think point 1) might not be that difficult to implement.  Possibly creating an Attribute name for every attribute to ensure we would never have an issue with special dynamo names in the filter expression.

For point 2), it appears that there is a GSI, but the HashKeyAttributeValue and HashKeyPropertyName are being set to "customerId" and the customerId value.  I'm looking for some input on what would be the right way to implement this feature and I wouldn't mind taking a shot at it myself.  I feel like there is somewhere else in the code base during the criteria creation where I would need to set something like a FilterKeyAttributeValue/PropertyName?

**Comment by derjust** [comment]
you are correct, an auto-filter feature is currently not in place. To fit into the Spring-Data world, a custom repository would be the way to go

**Comment by mswolfe** [comment]
@derjust Are you saying create a custom repository interface that defines a method like: findByCustomerIdAndEmail.  Then create an implementation that creates a new QueryRequest object and sends the request to DynamoDB?

I was hoping there was a way to utilize the AbstractDynamoDBQueryCreator/Criteria classes to turn a method name into a request with filters.  Something like, if the necessary hashkey/rangekey is present and a query can be created then do so and use filters for the remaining properties.  Is this approach too naive?
